src/librustc_mir/build/matches/mod.rs

```diff
@@ -937,13 +937,13 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
             for Binding { source, .. }
                 in matched_candidates.iter().flat_map(|candidate| &candidate.bindings)
             {
-                let mut cursor = &source.projection;
-                while let Some(box Projection { base, elem }) = cursor {
-                    cursor = base;
+                for (i, elem) in source.projection.iter().enumerate().rev() {
```

Comment:
Use `.find(...)` here? + `let proj_base = &source.projection[..i];` is only relevant if it is found.